src/main/java/scala/compat/java8/runtime/LambdaDeserializer.scala

```diff
@@ -0,0 +1,135 @@
+package scala.compat.java8.runtime
+
+import java.lang.invoke._
+import java.lang.ref.WeakReference
+
+/**
+ * This class is only intended to be called by synthetic `$deserializeLambda$` method that the Scala 2.12
+ * compiler will add to classes hosting lambdas.
+ *
+ * It is intended to be consumed directly.
+ */
+object LambdaDeserializer {
+  private final case class CacheKey(implClass: Class[_], implMethodName: String, implMethodSignature: String)
+  private val cache = new java.util.WeakHashMap[CacheKey, WeakReference[CallSite]]()
+
+  /**
+   * Deserialize a lambda by calling `LambdaMetafactory.altMetafactory` to spin up a lambda class
+   * and instantiating this class with the captured arguments.
+   *
+   * A cache is employed to ensure that subsequent deserialization of the same lambda expression
+   * is cheap, it amounts to a reflective call to the constructor of the previously created class.
+   * However, deserialization of the same lambda expression is not guaranteed to use the same class,
+   * concurrent deserialization of the same lambda expression may spin up more than one class.
+   *
+   * This cache is weak in keys and values to avoid retention of the enclosing class (and its classloader)
+   * of deserialized lambdas.
+   *
+   * Assumptions:
+   *  - No additional marker interfaces are required beyond `{java.io,scala.}Serializable`. These are
+   *    not stored in `SerializedLambda`, so we can't reconstitute them.
```

Comment:
One thing I still don't quite understand: why do we need to add the `Serializable` marker interface if the functional interface extends `Serializable` anyway? The created lambda object has a class type that implements the functional interface type, so it would have the right `Serializable` already.